### PR TITLE
Frontier Benchmarking (#453)

### PR DESCRIPTION
**Pull request by mohdsaid497566**
## Description
Added two benchmarking cases by submitting SLURM jobs on Frontier - duplicate implementation of Phoenix. (#453)

### CODE REVIEW

File: .github/workflows/frontier/bench.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+mkdir -p /storage/scratch1/6/sbryngelson3/mytmp_build
+export TMPDIR=/storage/scratch1/6/sbryngelson3/mytmp_build
+
+if [ "$job_device" == "gpu" ]; then
+    ./mfc.sh bench --mem 12 -j $n_ranks -o "$job_slug.yaml" -- -c frontier-bench $device_opts -n $n_ranks
+else
+    ./mfc.sh bench --mem 1 -j $(nproc) -o "$job_slug.yaml" -- -c frontier-bench $device_opts -n $n_ranks
+fi
+
+unset TMPDIR
```
Comment (sbryngelson) [inline .github/workflows/frontier/bench.sh]: remove

File: .github/workflows/frontier/bench.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+if [ "$job_device" == "gpu" ]; then
+    gpus=$(rocm-smi --showid | awk '{print $1}' | grep -Eo '[0-9]+' | uniq | tr '\n' ' ')
+    n_ranks=$(echo "$gpus" | wc -w)         # number of GPUs on node
+    gpu_ids=$(echo "$gpus" | tr ' ' '\n' | tr '\n' ' ' | sed 's/ $//')  # GPU IDs from rocm-smi
+    device_opts="--gpu -g $gpu_ids"
+fi
+
+mkdir -p /storage/scratch1/6/sbryngelson3/mytmp_build
+export TMPDIR=/storage/scratch1/6/sbryngelson3/mytmp_build
```
Comment (sbryngelson) [inline .github/workflows/frontier/bench.sh]: remove lines 12 and 13

File: .github/workflows/frontier/bench.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+if [ "$job_device" == "gpu" ]; then
+    ./mfc.sh bench --mem 12 -j $n_ranks -o "$job_slug.yaml" -- -c frontier-bench $device_opts -n $n_ranks
+else
+    ./mfc.sh bench --mem 1 -j $(nproc) -o "$job_slug.yaml" -- -c frontier-bench $device_opts -n $n_ranks
```
Comment (sbryngelson) [inline .github/workflows/frontier/bench.sh]: this `-c <template>` does not exist, nor does it need to exist. `-c frontier` is fine.

File: .github/workflows/bench.yml [hunk 4]
```diff
@@ -66,3 +66,58 @@ jobs:
             pr/build/benchmarks/*
             master/bench-${{ matrix.device }}.*
             master/build/benchmarks/*
+
+  frontier:
+      name: Oak Ridge | Frontier (AMD ROCm)
+      if: github.repository == 'MFlowCode/MFC' && needs.file-changes.outputs.checkall == 'true'
```
Comment (sbryngelson) [inline .github/workflows/bench.yml]: this almost entirely duplicates the code from phoenix... can you make a CI matrix instead to avoid the extra lines? also there is no need to measure performance on CPU cases on Frontier, that's just going to slow down the number of cases queued. i only have 1 frontier runner because that's the max num. of debug jobs one can submit.

Comment (Malmahrouqi3) [reply .github/workflows/bench.yml]: Yeah sure, I will apply those changes shortly.

File: .github/workflows/bench.yml [hunk 5]
```diff
@@ -19,15 +19,36 @@ jobs:
...
+            build_script: ""
+          - cluster: phoenix
+            name: Georgia Tech | Phoenix (NVHPC)
+            group: phoenix
+            labels: gt
+            flag: p
+            device: gpu
+            build_script: ""
+          - cluster: frontier
+            name: Oak Ridge | Frontier (AMD CCE)
```
Comment (sbryngelson) [inline .github/workflows/bench.yml]: just CCE

**Comment by Malmahrouqi3** [comment]
Reduced the job duration to 3 hrs to see whether it would yield the same error regardless of duration.

File: .github/workflows/frontier/bench.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+n_ranks=12
+
+if [ "$job_device" == "gpu" ]; then
+    gpus=$(rocm-smi --showid | awk '{print $1}' | grep -Eo '[0-9]+' | uniq | tr '\n' ' ')
+    n_ranks=$(echo "$gpus" | wc -w)         # number of GPUs on node
+    gpu_ids=$(echo "$gpus" | tr ' ' '\n' | tr '\n' ' ' | sed 's/ $//')  # GPU IDs from rocm-smi
+    device_opts="--gpu -g $gpu_ids"
+fi
+
+if [ "$job_device" == "gpu" ]; then
```
[the remark this answers is not in the document]
Comment (Malmahrouqi3) [reply .github/workflows/frontier/bench.sh]: Bash does not require explicit declaration.

File: .github/workflows/frontier/submit-bench.sh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+elif [ "$2" == "gpu" ]; then
+    sbatch_device_opts="\
+#SBATCH -n 8                       # Number of cores required"
+else
+    usage
+    exit 1
+fi
+
+
+job_slug="`basename "$1" | sed 's/\.sh$//' | sed 's/[^a-zA-Z0-9]/-/g'`-$2"
```
[the remark this answers is not in the document]
Comment (Malmahrouqi3) [reply .github/workflows/frontier/submit-bench.sh:27]: hmm, not helpful.

File: .github/workflows/frontier/bench.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (sbryngelson) [reply .github/workflows/frontier/bench.sh]: that isn't the point, is it?

**Comment by Malmahrouqi3** [comment]
Most Recent Failed Frontier-Benchmark Commit: https://github.com/MFlowCode/MFC/actions/runs/15743760213
``` bash
Run (cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  (cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  (cd master && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  wait %1 && wait %[2](https://github.com/MFlowCode/MFC/actions/runs/15743760213/job/44388664302#step:5:2)
  shell: /usr/bin/bash -e {0}
  env:
    ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION: node16
    ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION: true
bash: .github/workflows/frontier/submit-bench.sh: No such file or directory
sbatch: error: Batch script contains DOS line breaks (\r\n)
sbatch: error: instead of expected UNIX line breaks (\n).
Error: Process completed with exit code 1.
```

This PR has taken needlessly quite some time, so to bypass the wait time of the entire CI, I will remove all `.github/workflow` content/files except Frontier Benchmark test. I just wanted to confirm the existing bash scripts would success specifically for **[Oak Ridge | Frontier (CCE) (gpu)]**. Afterwards, offloaded files will be restored.


**Edit**: I tried `dos2unix` which did absolutely nothing, and when I `git add`'ed the bash files, they conceived no change at all.

**Review by sbryngelson — APPROVED** [review]

**Comment by sbryngelson** [comment]
> Most Recent Failed Frontier-Benchmark Commit: [MFlowCode/MFC/actions/runs/15743760213](https://github.com/MFlowCode/MFC/actions/runs/15743760213)
> 
> ```shell
> Run (cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
>   (cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
>   (cd master && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
>   wait %1 && wait %[2](https://github.com/MFlowCode/MFC/actions/runs/15743760213/job/44388664302#step:5:2)
>   shell: /usr/bin/bash -e {0}
>   env:
>     ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION: node16
>     ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION: true
> bash: .github/workflows/frontier/submit-bench.sh: No such file or directory
> sbatch: error: Batch script contains DOS line breaks (\r\n)
> sbatch: error: instead of expected UNIX line breaks (\n).
> Error: Process completed with exit code 1.
> ```
> 
> This PR has taken needlessly quite some time, so to bypass the wait time of the entire CI, I will remove all `.github/workflow` content/files except Frontier Benchmark test. I just wanted to confirm the existing bash scripts would success specifically for **[Oak Ridge | Frontier (CCE) (gpu)]**. Afterwards, offloaded files will be restored.
> 
> **Edit**: I tried `dos2unix` which did absolutely nothing, and when I `git add`'ed the bash files, they conceived no change at all.

I fixed it. You `dos2unix`ed the wrong file.

**Review by sbryngelson — APPROVED** [review]

**Comment by Malmahrouqi3** [comment]
I did `dos2unix` for all frontier directory files. Anyways thanks, I will wait if that is gonna pass the test now.

**Comment by sbryngelson** [comment]
This benchmark test will never pass in its current state because the Frontier files for benchmarking do not exist on the master branch, hence this error
```shell
(cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  (cd pr     && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  (cd master && bash .github/workflows/frontier/submit-bench.sh .github/workflows/frontier/bench.sh gpu) &
  wait %1 && wait %[2](https://github.com/MFlowCode/MFC/actions/runs/15826502985/job/44607985758?pr=881#step:5:2)
  shell: /usr/bin/bash -e {0}
  env:
    ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION: node16
    ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION: true
bash: .github/workflows/frontier/submit-bench.sh: No such file or directory
Submitted batch job [3](https://github.com/MFlowCode/MFC/actions/runs/15826502985/job/44607985758?pr=881#step:5:3)531713
```

once it looks like everything is working as well as one can expect, we can merge in the minimal files (`.github/workflows/*`) and then create a _new_ PR that tests it properly.

**Comment by Malmahrouqi3** [comment]
aight, myself or someone has to test it out manually by cloning master & pr and adding bash files in each then benchmarking on Frontier as a slurm/interative job to make sure nothing will corrupt in the process.

**Comment by wilfonba** [comment]
I verified that this works on my end. The IBM case still gives NaNs though...

**Comment by Malmahrouqi3** [comment]
> I verified that this works on my end. The IBM case still gives NaNs though...

Thanks much, and I wonder what the deal is with the IBM case ngl. Any specific error messages or such? If the issue persists, we can just exclude that case somehow. Also, NaNs I guess won't fail the test as can be seen on my recent PR when I assigned null to IBM grind/exec https://github.com/MFlowCode/MFC/pull/895#issuecomment-2992568747

**Edit**: lmk, if you suspect anything that might have caused that.

**Comment by wilfonba** [comment]
Well, the NaN issue was supposed to be fixed by #892 but it appears that that's not the case

**Comment by sbryngelson** [comment]
status?

**Comment by Malmahrouqi3** [comment]
@sbryngelson done on my end tbh and nothing to add

**Comment by sbryngelson** [comment]
> Well, the NaN issue was supposed to be fixed by #892 but it appears that that's not the case

what's going on here?

**Comment by wilfonba** [comment]
> > Well, the NaN issue was supposed to be fixed by #892 but it appears that that's not the case
> 
> what's going on here?

Any ideas @anandrdbz ?

**Comment by anandrdbz** [comment]
I'll look into it, last time I checked 2D_ibm was working, perhaps there were multiple issues causing NaNs

**Comment by anandrdbz** [comment]
I just ran 2D_ibm and 2D_ibm_multiphase to completion on an interactive node @wilfonba, is there another example case file that's failing ?

**Comment by sbryngelson** [comment]
> I just ran 2D_ibm and 2D_ibm_multiphase to completion on an interactive node @wilfonba, is there another example case file that's failing ?

It's the IBM case in the benchmarking cases (what this PR is about)

**Comment by anandrdbz** [comment]
Not sure when this was done but the case file in ibm in benchmarks does not actually have ib = T, in fact it's just running a single fluid hypo elastic case

**Comment by anandrdbz** [comment]
Also @Malmahrouqi3 I think you need to change the number of procs to 8 from 12. there are only 8GCDs per node on frontier so you might run into an error there

**Comment by anandrdbz** [comment]
Anyways, I believe the reason why this particular case fails obviously has nothing to do with IBM since ib is not set, I think the reason is the problem size on frontier is larger than Phoenix due to it using 8 GPUs while the time step is hardcoded. I ran the same case file on a single GCD on frontier and it worked. I also reduced dt by a factor of 2 on 8 ranks and that also runs.  

But I guess there's not much point debugging this since there needs to be an overhaul of the case file to include an actual IBM case

**Comment by sbryngelson** [comment]
waiting for CI to run them will merge

**Review by sbryngelson — APPROVED** [review]

**Review by sbryngelson — APPROVED** [review]
